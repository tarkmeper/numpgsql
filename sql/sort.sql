-- Basic sorting
SELECT sort('{}'::int[]);
SELECT sort('{1}'::int[]);
SELECT sort('{5,1,4,6}'::int[]);

--Support Types
SELECT sort('{5,1,4,6}'::smallint[]);
SELECT sort('{5,1,4,6}'::bigint[]);
SELECT sort('{5,1,4,6}'::real[]);
SELECT sort('{5,1,4,6}'::double precision[]);

-- Unsupported type
SELECT SORT(ARRAY[true,true,false]);

-- Check reverse
SELECT reverse('{}'::int[]);
SELECT reverse('{1}'::int[]);
SELECT reverse('{5,1,4,6}'::int[]);

-- Check other types
SELECT reverse(ARRAY[true,true,false]);
SELECT reverse(ARRAY['cat', 'dog', 'rat']);

--multi dimesnion 
SELECT reverse('{{5,1},{4,6}}'::int[]);
