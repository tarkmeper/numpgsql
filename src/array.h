extern "C"  {
        #include "postgres.h"
        #include "catalog/pg_type.h"
        #include "utils/array.h"
        #include "fmgr.h"
}
#include <vector>
#include <tuple>


inline static size_t arr_cnt(const ArrayType* arr) {
	size_t result = 1;
	for (int i = 0; i < ARR_NDIM(arr); ++i){
		result *= ARR_DIMS(arr)[i];
	}
	return result;
}

//Return values - TODO rename function this was poorly named.
template<class T> inline Datum _ret(const T& val) { T::unimplemented_function; return Datum(); }
template<> inline Datum _ret<double>(const double& val) { PG_RETURN_FLOAT8(val); }
template<> inline Datum _ret<float>(const float& val) { PG_RETURN_FLOAT4(val); }
template<> inline Datum _ret<short>(const short& val) { PG_RETURN_INT16(val); }
template<> inline Datum _ret<int>(const int& val) { PG_RETURN_INT32(val); }
template<> inline Datum _ret<long>(const long& val) { PG_RETURN_INT64(val); }
template<> inline Datum _ret<bool>(const bool& val) { PG_RETURN_BOOL(val); }

//OID types
template<class T> inline Oid oid_type() { return T::unimplemented_feature; }
template<> inline Oid oid_type<float>() { return FLOAT4OID; }
template<> inline Oid oid_type<double>() { return FLOAT8OID; }
template<> inline Oid oid_type<short>() { return INT2OID; }
template<> inline Oid oid_type<int>() { return INT4OID; }
template<> inline Oid oid_type<long>() { return INT8OID; }
template<> inline Oid oid_type<bool>() { return BOOLOID; }

//Alignment codes
template<class T> char oid_align() { return T::unimplemented_feature; }
template<> inline char oid_align<float>() { return 'i'; }
template<> inline char oid_align<double>() { return 'd'; }
template<> inline char oid_align<short>() { return 's'; }
template<> inline char oid_align<int>() { return 'i'; }
template<> inline char oid_align<long>() { return 'd'; }
template<> inline char oid_align<bool>() { return 'c'; }

//Get param values
template<class T> inline T get_param(PG_FUNCTION_ARGS, int p)  { T::unimplemented_function; }
template<> inline bool get_param<bool>(PG_FUNCTION_ARGS, int p) { return PG_GETARG_BOOL(p); }
template<> inline float get_param<float>(PG_FUNCTION_ARGS, int p) { return PG_GETARG_FLOAT4(p); }
template<> inline double get_param<double>(PG_FUNCTION_ARGS, int p) { return PG_GETARG_FLOAT8(p); }
template<> inline short get_param<short>(PG_FUNCTION_ARGS, int p) { return PG_GETARG_INT16(p); }
template<> inline int get_param<int>(PG_FUNCTION_ARGS, int p) { return PG_GETARG_INT32(p); }
template<> inline long get_param<long>(PG_FUNCTION_ARGS, int p) { return PG_GETARG_INT64(p); }
