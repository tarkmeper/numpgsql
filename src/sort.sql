-- This program is open source, licensed under the MIT License.
-- For license terms, see the LICENSE file.
--
-- Copyright (C) 2019: Mark Teper

-- Other functions
CREATE FUNCTION sort(anyarray) RETURNS anyarray LANGUAGE C STRICT IMMUTABLE LEAKPROOF COST 100 PARALLEL SAFE AS '$libdir/numpgsql', 'sort';
CREATE FUNCTION reverse(anyarray) RETURNS anyarray LANGUAGE C STRICT IMMUTABLE LEAKPROOF COST 100 PARALLEL SAFE AS '$libdir/numpgsql', 'reverse';
