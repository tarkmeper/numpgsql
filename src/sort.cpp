extern "C"  {
	#include "postgres.h"
	#include "utils/array.h"
	#include "fmgr.h"
        #include <utils/lsyscache.h>

	PG_FUNCTION_INFO_V1(sort); Datum sort(PG_FUNCTION_ARGS);
	PG_FUNCTION_INFO_V1(reverse); Datum reverse(PG_FUNCTION_ARGS);
}
#include "array.h"

#include <algorithm>
#include <tuple>

template<class T> void do_sort(void* v, const size_t n) {
    T* f = reinterpret_cast<T*>(v);
    std::sort(f, f + n );
}

/***** operator template ****/
Datum sort(PG_FUNCTION_ARGS) {
	ArrayType* arr = PG_GETARG_ARRAYTYPE_P_COPY(0);

        const size_t n = arr_cnt(arr);
        if (n == 0 ) { PG_RETURN_POINTER(arr); }

        Oid type = ARR_ELEMTYPE(arr);
        void * v = ARR_DATA_PTR(arr);

        switch(type) {
            case FLOAT4OID: do_sort< float > (v, n); break;
            case FLOAT8OID: do_sort< double > (v, n); break;
            case INT2OID: do_sort< int16 > (v, n); break;
            case INT4OID: do_sort< int32 > (v, n); break;
            case INT8OID: do_sort< int64 > (v, n); break;
            default:
                elog(ERROR, "Unsupported array type");
                PG_RETURN_NULL();
        }
	PG_RETURN_POINTER(arr);
}


//There may be a slightly better version that can do this swap in place.
Datum reverse(PG_FUNCTION_ARGS) {
  ArrayType* arr = PG_GETARG_ARRAYTYPE_P(0);

  Oid        type  = ARR_ELEMTYPE(arr);
  int16      elmlen;
  bool       elmbyval;
  char       elmalign;
  Datum*     elemsp;
  bool*      nullsp;
  int        nelemsp;

  if (ARR_NDIM(arr) == 0 ) {
     PG_RETURN_POINTER(construct_empty_array(type));
  } else if (ARR_NDIM(arr) != 1 ) {
     elog(ERROR, "Currently reverse only support 1d arrays");
     PG_RETURN_NULL();
  }

  const size_t n = ARR_DIMS(arr)[0];
  get_typlenbyvalalign(type, &elmlen, &elmbyval, &elmalign);
  deconstruct_array(arr, type, elmlen, elmbyval, elmalign, &elemsp, &nullsp, &nelemsp);

  Datum* rev = (Datum*)palloc(n * sizeof(Datum));
  std::copy(elemsp, elemsp + n, rev);
  std::reverse(rev, rev + n);

  ArrayType* res = construct_array(rev, n, type, elmlen, elmbyval, elmalign);
  PG_RETURN_POINTER(res);
}
