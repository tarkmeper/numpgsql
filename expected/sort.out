-- Basic sorting
SELECT sort('{}'::int[]);
 sort 
------
 {}
(1 row)

SELECT sort('{1}'::int[]);
 sort 
------
 {1}
(1 row)

SELECT sort('{5,1,4,6}'::int[]);
   sort    
-----------
 {1,4,5,6}
(1 row)

--Support Types
SELECT sort('{5,1,4,6}'::smallint[]);
   sort    
-----------
 {1,4,5,6}
(1 row)

SELECT sort('{5,1,4,6}'::bigint[]);
   sort    
-----------
 {1,4,5,6}
(1 row)

SELECT sort('{5,1,4,6}'::real[]);
   sort    
-----------
 {1,4,5,6}
(1 row)

SELECT sort('{5,1,4,6}'::double precision[]);
   sort    
-----------
 {1,4,5,6}
(1 row)

-- Unsupported type
SELECT SORT(ARRAY[true,true,false]);
ERROR:  Unsupported array type
-- Check reverse
SELECT reverse('{}'::int[]);
 reverse 
---------
 {}
(1 row)

SELECT reverse('{1}'::int[]);
 reverse 
---------
 {1}
(1 row)

SELECT reverse('{5,1,4,6}'::int[]);
  reverse  
-----------
 {6,4,1,5}
(1 row)

-- Check other types
SELECT reverse(ARRAY[true,true,false]);
 reverse 
---------
 {f,t,t}
(1 row)

SELECT reverse(ARRAY['cat', 'dog', 'rat']);
    reverse    
---------------
 {rat,dog,cat}
(1 row)

--multi dimesnion 
SELECT reverse('{{5,1},{4,6}}'::int[]);
ERROR:  Currently reverse only support 1d arrays
